internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go

```diff
@@ -0,0 +1,318 @@
+var _ dynamicScaleDownExecutor = &standardDynamicScaleDownProcessor{}
+var _ dynamicScaleDownProcessor = &standardDynamicScaleDownProcessor{}
+
+// ShouldScaleDown indicates whether a new data plane
+// cluster should be created for a given instance type in the given provider
+// and region.
+// It returns true if all the following conditions happen:
+// 1. If specified the cluster is empty i.e it does not contain any streaming unit
+// 2. If the cluster can be removed without triggering a scale up action
+// Otherwise false is returned.
+// Note: This method assumes kafkaStreamingUnitCountPerClusterList does not
+// contain elements with the Status attribute with the 'failed' value.
+// Thus, if the type is constructed with the assumptions being true, it
+// can be considered as the 'failed' state Clusters are not included in
+// the calculations.
+func (p *standardDynamicScaleDownProcessor) ShouldScaleDown() (bool, error) {
+	// First let's check if the cluster is empty
+	for _, i := range p.indexesOfStreamingUnitForSameClusterID {
+		clusterIsNotEmpty := p.kafkaStreamingUnitCountPerClusterList[i].Count > 0
+		if clusterIsNotEmpty {
+			return false, nil
+		}
+	}
+
+	// let's check if the cluster can be safely removed without causing a scale up event
+	if len(p.regionsSupportedInstanceType) == 0 { // if no region limits are available it means that this cluster is in a region that's not supported anymore, we can safely delete it if it is empty
+		return true, nil
+	}
+
+	newkafkaStreamingUnitCountPerClusterList := services.KafkaStreamingUnitCountPerClusterList{}
+
+	clusterStatesTowardReadyState := []string{
+		api.ClusterProvisioning.String(), api.ClusterProvisioned.String(),
+		api.ClusterAccepted.String(), api.ClusterWaitingForKasFleetShardOperator.String(),
+	}
+
+	for _, suCount := range p.kafkaStreamingUnitCountPerClusterList {
+		// Ignore clusters in terraforming states for scale up evaluation as they are not ready yet.
+		// We only want to delete the cluster if we've a sibling ready cluster
+		if arrays.Contains(clusterStatesTowardReadyState, suCount.Status) {
+			continue
+		}
+
+		// Keep only clusters streaming unit count for cluster that are not going to be deleted.
+		// i.e Assume that the candidate cluster is removed from the new list which will be used to perform scale up evaluation
+		if suCount.ClusterId == p.clusterID {
+			continue
+		}
+
+		newkafkaStreamingUnitCountPerClusterList = append(newkafkaStreamingUnitCountPerClusterList, suCount)
+	}
+
+	for _, i := range p.indexesOfStreamingUnitForSameClusterID {
+		suCount := p.kafkaStreamingUnitCountPerClusterList[i]
+
+		currLocator := supportedInstanceTypeLocator{
+			provider:         suCount.CloudProvider,
+			region:           suCount.Region,
+			instanceTypeName: suCount.InstanceType,
+		}
+
+		instanceTypeConfig, ok := p.regionsSupportedInstanceType[suCount.InstanceType]
+		if !ok {
+			continue
+		}
+
+		dynamicScaleUpProcessor := &standardDynamicScaleUpProcessor{
```

Comment:
I am thinking if using the dynamic scale up processor is reliable for this purpose the way it is implemented in this PR now:
One of the conditions to scale down is that the number of allocatable kafka instances in the region would still be higher than the defined capacity slacks for **all** supported instance types if the cluster were to be deleted.
However the dynamic scale up processor only evaluates it for an individual instance type.
What happens if the cluster being evaluated supports multiple instance types?

Reply:
Also are we redundantly calculating it? The list is at kafka size level and not at kafka instance type level

Reply:
That's a great question. 

We evaluate scale up for all the supported instance types of the candidate cluster.
The logic in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/6c071fb6699f56cbc3dc379abdaf1c9a9fa15df8/internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go#L259-L291 handles the multiple instance type cases

Reply:
Not only multiple instance types, but instance types **and** instance sizes of that type no?

Reply:
Yes. That's already handled in that loop.

Reply:
But what I mean is that we are redundantly calculating the shouldScaleUp no? because shouldScaleUp is at instance type level. It considers all sizes of the instance type

Reply:
Thanks for the clarification. 

> But what I mean is that we are redundantly calculating the shouldScaleUp no? because shouldScaleUp is at instance type level. It considers all sizes of the instance type.

No. Because, here we are dealing only with instance type. The largest size consideration (which I initially wrongly thought was your question) is handled in the `.ShouldScaleUp()` call. 
So at the end of the loop, for each instance type, we'll call `.ShouldScaleUp()` once.

> For example if you have instance type standard with 6 different sizes you only need to call shouldScaleUp once.  However the current logic calls it 6 times

For each instance type we'll call `.ShouldScaleUp()` once. In the `kafkaStreamingUnitCountPerClusterList` list, an instance type is only represented once. All the different sizes are agregated in `.Count` field of each element in that list.

Reply:
> No. Because, here we are dealing only with instance type.

Is that correct? I understand that if we have an instance type 'standard' with 2 sizes then wouldn't `kafkaStreamingUnitCountPerClusterList` have 2 entries and thus shouldScaleUp be called for each one?

Reply:
For the same `cluster_id` then there will be one entry and the sizes as considered/treated in the `.Count` - https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/8aa0a7614ce607c101f0ef85533fb8f9684c7b13/internal/kafka/internal/services/clusters.go#L799

Reply:
I see. I missed that part. It seems correct then :+1: